### PR TITLE
[Merged by Bors] - Refine compaction

### PR DESCRIPTION
**Pull request by michaelsproul**
## Proposed Changes

In an attempt to fix OOM issues and database consistency issues observed by some users after the introduction of compaction in v0.3.4, this PR makes the following changes:

* Run compaction less often: roughly every 1024 epochs, including after long periods of non-finality. I think the division check proposed by Paul is pretty solid, and ensures we don't miss any events where we should be compacting. LevelDB lacks an easy way to check the size of the DB, which would be another good trigger.
* Make it possible to disable the compaction on finalization using `--auto-compact-db=false`
* Make it possible to trigger a manual, single-threaded foreground compaction on start-up using `--compact-db`
* Downgrade the pruning log to `DEBUG`, as it's particularly noisy during sync

I would like to ship these changes to affected users ASAP, and will document them further in the Advanced Database section of the book if they prove effective.

### CODE REVIEW

File: beacon_node/src/cli.rs [hunk 1]
```diff
@@ -297,13 +297,25 @@ pub fn cli_app<'a, 'b>() -> App<'a, 'b> {
...
          */
         .arg(
             Arg::with_name("purge-db")
                 .long("purge-db")
                 .help("If present, the chain database will be deleted. Use with caution.")
         )
+        .arg(
+            Arg::with_name("compact-db")
+                .long("compact-db")
+                .help("If present, apply compaction to the database on start-up. Use with caution.")
```
Comment (paulhauner) [inline beacon_node/src/cli.rs:310]:
```suggestion
                .help("If present, apply compaction to the database on start-up. Use with caution, it is generally not recommended unless auto-compaction is disabled.")
```

File: beacon_node/beacon_chain/src/migrate.rs [hunk 2]
```diff
@@ -203,15 +211,25 @@ impl<E: EthSpec, Hot: ItemStore<E>, Cold: ItemStore<E>> BackgroundMigrator<E, Ho
...
-                log,
-                "Database compaction failed";
-                "error" => format!("{:?}", e)
-            );
+        // We use a heuristic to avoid compacting too often, as it's potentially costly and puts
+        // wear on the disk.
+        let old_finalized_epoch = old_finalized_checkpoint.epoch;
+        let new_finalized_epoch = notif.finalized_checkpoint.epoch;
+        if db.compact_on_prune()
+            && (new_finalized_epoch % COMPACTION_MODULO_THRESHOLD == 0
```
Comment (paulhauner) [inline beacon_node/beacon_chain/src/migrate.rs:219]: What are your thoughts on this:

```rust
old_finalized_epoch \ COMPACTION_MODULO_THRESHOLD < new_finalized_epoch \ COMPACTION_MODULO_THRESHOLD
```

This way we also detect when finality crosses a `COMPACTION_MODULO_THRESHOLD` (i.e., we detect when we skip-finalize a threshold) and we don't need the checks on the following line.

File: beacon_node/src/cli.rs (hunk 1 above)
Comment (paulhauner) [reply beacon_node/src/cli.rs:310]: Suggestion, feel free to disregard :)

**Review by paulhauner — CHANGES_REQUESTED** [review]
Looks good, only two comments/suggestions.

File: beacon_node/beacon_chain/src/migrate.rs (hunk 2 above)
Comment (michaelsproul) [reply beacon_node/beacon_chain/src/migrate.rs:219]: Yes!! Perfect!

**Comment by michaelsproul** [comment]
I'm going to test sync on an HDD to see if this resolves issues, but I'm also happy to merge this in the meantime, as I think all these improvements are generally helpful, even if they don't 100% solve the OOM issue.

**Review by paulhauner — APPROVED** [review]

**Comment by michaelsproul** [comment]
on second thoughts...

bors r-

I want to do some more testing. I suspect based on Dankrad's reports that the compaction is still too frequent during sync (23 compactions during Medalla sync is still a lot): https://github.com/sigp/lighthouse/issues/1917#issuecomment-728464640

**Review by paulhauner — APPROVED** [review]
Looks good! We spent a fair bit of time chatting about this out-of-band and I think the present compaction timing is good for now.

Merge it!

**Pull request closed without merging**